### PR TITLE
Adding the QR Code feature

### PR DESCRIPTION
**Pull request by macintosh1398**
Adding the first draft of a QR Code reader. A new QR-Code reader activity was created, launched by a button on the main activity, that is able to read and display the result of a QR-Code scan as a "Toast" text.

### CODE REVIEW

File: app/src/androidTest/java/com/github/multimatum_team/multimatum/QRCodeReaderActivityTest.kt [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import org.junit.Test
+
+class QRCodeReaderActivityTest {
+    @get:Rule
+    val permissionRule: GrantPermissionRule = GrantPermissionRule.grant(
+        android.Manifest.permission.CAMERA
+    )
+
+    @Test
+    fun should_display_scanner_when_permissionsAreGranted() {
```
Comment (felko) [inline app/src/androidTest/java/com/github/multimatum_team/multimatum/QRCodeReaderActivityTest.kt:18]: You can use this syntax to explain the test scenario in plain english:
```kotlin
fun `Should display scanner when permissions are granted`() {
    ...
}
```

File: app/src/main/java/com/github/multimatum_team/multimatum/MainActivity.kt [hunk 2]
```diff
@@ -1,10 +1,13 @@
 package com.github.multimatum_team.multimatum
 
 import android.content.Intent
+import android.content.pm.PackageManager
 import androidx.appcompat.app.AppCompatActivity
 import android.os.Bundle
 import android.view.View
 import android.widget.EditText
```
Comment (felko) [inline app/src/main/java/com/github/multimatum_team/multimatum/MainActivity.kt:8]: I think this import is not used

File: app/src/main/java/com/github/multimatum_team/multimatum/QRCodeReaderActivity.kt [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import android.os.Bundle
+import android.widget.Toast
+import com.budiyev.android.codescanner.AutoFocusMode
+import com.budiyev.android.codescanner.CodeScanner
+import com.budiyev.android.codescanner.CodeScannerView
+import com.budiyev.android.codescanner.DecodeCallback
+import com.budiyev.android.codescanner.ErrorCallback
+import com.budiyev.android.codescanner.ScanMode
+
+class QRCodeReaderActivity : AppCompatActivity() {
```
Comment (felko) [inline app/src/main/java/com/github/multimatum_team/multimatum/QRCodeReaderActivity.kt:13]: Would be nice to have a brief description of the activity

**Review by felko — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/github/multimatum_team/multimatum/MainActivity.kt (hunk 2 above)
Comment (felko) [reply app/src/main/java/com/github/multimatum_team/multimatum/MainActivity.kt:8]: Sorry wrong line, I meant the `import android.widget.EditText` import

File: app/src/androidTest/java/com/github/multimatum_team/multimatum/QRCodeReaderActivityTest.kt (hunk 1 above)
Comment (ValentinAebi) [reply app/src/androidTest/java/com/github/multimatum_team/multimatum/QRCodeReaderActivityTest.kt:18]: @felko maybe be careful with this "plain english" syntax. I have already had problems with spaces not recognized and replaced by codes in error messages in this syntax, which is unpleasant to read. I would leave it with "_" (but it is a detail).

File: app/src/main/java/com/github/multimatum_team/multimatum/QRCodeReaderActivity.kt [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                generateCallbackTest("Camera initialization error: ${it.message}")
+            }
+        }
+
+        // Start the scanner
+        scannerView.setOnClickListener {
+            codeScanner.startPreview()
+        }
+    }
+
```
Comment (ValentinAebi) [inline app/src/main/java/com/github/multimatum_team/multimatum/QRCodeReaderActivity.kt:51]: The name of this method looks a bit strange. Would it not be less confusing to name it e.g. `displayOnToast`?

File: app/src/main/java/com/github/multimatum_team/multimatum/QRCodeReaderActivity.kt [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    fun generateCallbackTest(text: String){
+        Toast.makeText(this, text, Toast.LENGTH_LONG).show()
```
Comment (ValentinAebi) [inline app/src/main/java/com/github/multimatum_team/multimatum/QRCodeReaderActivity.kt:53]: The name of this method looks a bit strange. Would it not be less confusing to name it e.g. `displayOnToast`?

File: app/src/androidTest/java/com/github/multimatum_team/multimatum/QRCodeReaderActivityTest.kt (hunk 1 above)
Comment (macintosh1398) [reply app/src/androidTest/java/com/github/multimatum_team/multimatum/QRCodeReaderActivityTest.kt:18]: For now I will let it like that. Maybe change in the future

**Review by felko — APPROVED** [review]
I approve even though the coverage has decreased a lot, we expect that roboelectric will let us write more tests for the camera

**Comment by macintosh1398** [comment]
The QR-Code reader works and can be launched using a button on the main screen. Manual tests were also done to check the ability of the camera to scan QR-Codes. I have improved the coverage from 40% to 87% this week, which is a lot better. For now I don't think I can create more tests but in the future we'll see (more comfortable with UI tests).
